### PR TITLE
Event log blog post

### PR DESCRIPTION
**Pull request by tsx**

### CODE REVIEW

File: _posts/2022-01-05-event-log/index.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+We expose Event Log in two ways to our customers: Event Log API and Webhooks. We’re also using the same Event Log internally for search and indexing implementation.
+
+# Event Log API
+
+Event Log API is a REST API that lets you retrieve events that happened within past 30 days.
+
+Historical events are stored in a MongoDB cluster that has a [TTL](https://docs.mongodb.com/manual/core/index-ttl/) set on events collection. Once you have all the events in a database, serving them with a typical [REST API](https://developer.close.com/resources/event-log/list-of-events/) is very straightforward. What’s not so straightforward is getting them *into* said database.
+
+For quite a while, we used Redis as an intermediate queue between the application emitting events and the service that exposes them. In our main application code, we’d serialize an event to a JSON string, then push the new event into a Redis list which works as a queue. On Events Service side, there are “collector” processes that take a batch of events from Redis and save them into MongoDB database after some light preprocessing.
```
Comment (wojcikstefan) [inline _posts/2022-01-05-event-log/index.md:24]: I think "which worked as a queue" would be more correct than "which works as a queue"

**Review by wojcikstefan — APPROVED** [review]
👏 🚀

File: _posts/2022-01-05-event-log/index.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+# Kafka
+
+Using Redis as an intermediate queue worked pretty well for us for quite some time. However as our customer base, number of events, and reliance on Event Log for production-critical systems grew, we started hitting limits of this architecture. Here are a few concerns we had:
+
+- During normal operation, the lists that work as a queue are near-empty. But if receiving side goes down for any reason and stops removing events from lists, the queue starts growing. Redis, being an in-memory storage, can only hold so much data. Our estimates were that we would only have a few hours worth of space before we would need to drop events if the receiving side had a prolonged outage.
+- The Redis instance serving the queue is itself a single point of failure. While we’re using ElastiCache that offers failover support, it is not entirely seamless and still has potential to lose some events that were written to primary but did not propagate to a secondary instance in a failure event.
+- This architecture does not make it easy to have multiple independent consumers. To use Event Log in an indexing pipeline, we had to have a second Redis instance and send all events to both instances.
```
Comment (jkemp101) [inline _posts/2022-01-05-event-log/index.md:46]: >To use Event Log in an indexing pipeline, we had to have a second Redis instance and send all events to both instances.

This transition seems abrupt and took me a second to confirm you were introducing a new consumer, the indexing pipeline. Maybe reword to something like:

>For instance, when we wanted our new indexing pipeline to use these same events, we had to add another Redis instance and send all events to both instances.

File: _posts/2022-01-05-event-log/index.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+- Implementing a robust retry strategy was a challenge. Since we pop an item from a list for processing, we’ll have to re-insert it somewhere else to make another attempt at processing it later in case of a failure.
+
+About a year ago, we moved Event Log queue to [Kafka](https://kafka.apache.org) instead of Redis. Kafka actually stores all events as an append-only log for a set amount of time, and consumers are represented as pointers or positions within the log. This makes it a lot easier to scale, support multiple consumers from the same Kafka topic, and handle failures. As a bonus, we gained an ability to replay event stream from some point in the past in an unlikely case if something goes really wrong with any consumer.
+
+We made a zero-downtime switch from Redis to Kafka by first improving consumers to handle duplicate events correctly and process them in idempotent way. Then we started sending events to both queues at the same time while closely monitoring metrics of both queues. After we ensured correct operation of the Kafka-based queue, we stopped sending events to Redis, and followed up with removing Redis-based code.
```
Comment (jkemp101) [inline _posts/2022-01-05-event-log/index.md:51]: >Then we started sending events to both queues

Technically we were already sending to both "queues", did you mean then we started processing them from both queues?

File: _posts/2022-01-05-event-log/index.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+From a more technical perspective, there are multiple ways to create events. An early approach we’ve taken to bootstrap event log is to monkey-patch MongoEngine document classes to report previous (received from a `findAndModify` [mongo operation](https://docs.mongodb.com/manual/reference/command/findAndModify/)) and new data, and format that to an event using MongoRest resources. This allowed us to add events to almost all object types essentially for free, reusing existing API-compatible code we already had. Providing events in a shape that matches existing API was a really good way to ensure consistency.
```
Comment (jkemp101) [inline _posts/2022-01-05-event-log/index.md:100]: I would link `MongoRest` to https://github.com/closeio/flask-mongorest

File: _posts/2022-01-05-event-log/index.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        Action.Updated,
+        previous_data=previous_data,
+        data=serialize_pipeline_event(pipeline),
+    )
+    return pipeline
+```
+
+# Conclusion
+
+Making integrations with Close easy was always a priority for our Engineering team. On top of providing [world-class REST API support](https://developer.close.com), Event Log and Webhooks are indispensable tools for building integrations that react to user activity in Close app. We’re looking forward to integrating with more tools in the ecosystem.
```
Comment (jkemp101) [inline _posts/2022-01-05-event-log/index.md:123]: Probably not important enough to add to this post unless you think so but the only other factor I can think of is how we need to control how some events are processed, like running a migration and not wanting to trigger webhooks to be sent to customers. Seems a little interesting because it is the kind of thing you think of after the fact.

**Review by jkemp101 — APPROVED** [review]

File: _posts/2022-01-05-event-log/index.md (hunk 5 above)
Comment (tsx) [reply _posts/2022-01-05-event-log/index.md:142]: Great point. Added a section about that.
